web: (RFC14) Add invalid keyword highlighting

hadrian-git opened the pull request:
**Changes:**

- Invalid search keywords are now decorated with a red wavy underline
- The search input has been refactored to use a `<div contentEditable>`
  - Enables rendering of styled text in the search input

![rfc14](https://user-images.githubusercontent.com/1495145/70101037-9fe1c680-162b-11ea-82d7-a06565608379.gif)

**Implements:**

- [RFC 14 APPROVED: Improve autocomplete suggestions](https://drive.google.com/open?id=1vSzZayA6p-znJ-7y8XrdlzluvZhboYytEuQ465_TTM4)

**Notes:**

- The invalid keyword tags are rendered using a HTML string. It's possible to use React elements with the attribute `suppressContentEditableWarning`, but that will require a refactor of the cursor position calculation, including how suggestions are added to the query, as that was coded for a single text `<input>`. I'd like to leave this for a future PR so we can have time to discuss on the next feature to the search input and how that can be developed. The current additions are well isolated, and if it doesn't work out, the components can easily be switched back to using a single `<input>`
- Initial implementation informs of the invalid keyword using the `title` attribute. Future implementations may show quick-fix actions, but that will need some extra planning (previous item on refactoring)
- I'm planning on a PR to split out the functions in `web/src/search/helpers.tsx`, the file is getting a bit long and might cause unnecessary conflicts on future code updates
- I tried out [Draft.js](https://draftjs.org/) but couldn't find my way around getting it to render correctly with our styling, and get it to work correctly in a single line (also using plug-ins). But now with the contentEditable rendering and working correctly, maybe we can revisit Draft.js for 3.12
- While updating the regex I noticed that [RegExr](https://regexr.com/4qgpg) is using [CodeMirror](https://codemirror.net/) for its styled single line input. CodeMirror can also be an option for our use case

**Updates:**

- Dec/12/2019: Investigating usage of CodeMirror as a replacement to `ContentEditableInput`

rvantonder commented:
Cool feature! Two things I noticed:

1. The regex doesn't detect quoted search patterns, so we will highlight things like this:

![Screen Shot 2019-12-04 at 12 58 11 AM](https://user-images.githubusercontent.com/888624/70124205-77f56180-1631-11ea-85d5-6ab389760b2b.png)

I actually would not advocate for a regex approach to solve this, it is brittle (single and double quotes need to be considered, but it is dependent on the `patterntype` in effect, etc.).

2. The `patterntype` keyword is not recognized:

![Screen Shot 2019-12-04 at 12 58 44 AM](https://user-images.githubusercontent.com/888624/70124277-aa9f5a00-1631-11ea-8111-4caf2f10a909.png)

Ideally both these issues would be handled by [RFC 75](https://docs.google.com/document/d/1iAWIOalYTZLRWmYw0r6_r_uThutLhanyTrxFSFtQAgE/edit#)(the backend would recognize what is and isn't a keyword, and tell you which keywords are invalid, and their position). But that's still work in progress. I realize we may want to ship this highlight feature before a that workable solution exists, so for (1) I'm inclined to just recommend punting on trying to do this better with a regex pattern.

hadrian-git commented:
@rvantonder thanks for the feedback! Indeed, RFC 75 looks like a better solution. For the `patterntype` keyword, I'll add it to the list of valid keywords. For the regex, I'll look into updating it to not match inside quotes.

lguychard commented:
On the homepage query field, the squiggles are truncated:

![image](https://user-images.githubusercontent.com/1741180/70165617-37c5cb80-1691-11ea-9dd2-e29901e57762.png)

In the "main" query field, their display is a bit glitchy as you type, and they don't appear to extend to the full length of the invalid keyword:

![squiggles](https://user-images.githubusercontent.com/1741180/70166020-ed911a00-1691-11ea-8887-825df54e4d2f.gif)

Could this be improved?

felixfbecker commented:
One problem with `contenteditable` is that it allows rich content, so my default I could paste e.g. bold text into it, or even dangerous content containing scripts or event handlers. I just wanted to test this, but pasting actually doesn't work at all. Console output after pasting:

![image](https://user-images.githubusercontent.com/10532611/70175964-802eba80-168c-11ea-8b6d-fd69fb6429b9.png)

attfarhan commented:
@felixfbecker I am able to paste normal text/bolded text. Are you trying to paste GIFs or something else?

felixfbecker commented:
I am pasting bold text copied from Microsoft Word

attfarhan commented:
I was able to copy-paste some bold text from the Notes app, but it results in a very odd query. 
Because content-editable allows rich content to be inputted, I imagine we may run into many issues like this: 

![2019-12-04 17 25 49](https://user-images.githubusercontent.com/16265452/70195690-36a99400-16bb-11ea-848f-18bcaf16d56e.gif)

We may have to sanitize the input somehow for this to make sense.

hadrian-git commented:
@attfarhan @felixfbecker You can now try to paste in rich text. If it does not render to plain text, please ping me so I can get the content you pasted, and debug it locally.

@rvantonder I've updated the regex to exclude keywords in quotes.

@lguychard The broken rendering seems to be an issue with chromium based browsers. I've updated the styling to an underline.

hadrian-git commented:
> This PR should include a CHANGELOG entry.
> 
> I feel like dashed would look a bit better than just a solid underline. Wdyt?
> 
> image
> 
> More generally: I see a lot of tricky code handling edge cases of contenteditable in React, eg.:
> 
>     DOM mutation in a requestAnimationFrame callback.
>     Non-trivial queryStringCursorPosition(), manipulating raw DOM elements.
>     Impure setCursor() calling document.getSelection() and manipulating raw DOM elements.
> 
> While AFAICT the functionality here looks good, the previously reported bugs on rich content handling, and the aforementioned complexity of the code, make me a bit wary of introducing this hand-rolled wrapper in one of our most critical UI components. This type of complexity is typically best offloaded to a third-party component (where edge cases have been battle-tested by many library users over time). You mention having discovered yesterday that regexr uses CodeMirror to implement a similar input element. Did you previously (or recently) weigh CodeMirror against DraftJS, or investigate how suitable CodeMirror would be here? I'm also curious to get the intuition of more @sourcegraph/web members here.

@lguychard I'm ok with both the underline and dashed styling. For the component, I'm currently investigating how CodeMirror would suite this case. Monaco might also work if it has an non-hacky way to render as a single-line input.

hadrian-git commented on web/src/search/input/ContentEditableInput.tsx:
```diff
@@ -0,0 +1,265 @@
+ * Content values for `ContentEditableInput` component.
+ * See `ContentEditableInput->Props['value']`
+ */
+export class ContentEditableState {
```
Forgot to update this one. I'll look into it after concluding on the CodeMirror usage

beyang commented:
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.11 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you
